File: shared/python/error_reporter.py
"""Shared error reporter — services use this to push errors to audit-service."""
from __future__ import annotations

import logging
import os
import socket
import traceback
from typing import Any

import httpx

logger = logging.getLogger("error_reporter")


class ErrorReporter:
    """Report errors to centralized audit-service.

    Usage:
        from shared.observability.error_reporter import report_error

        try:
            risky_op()
        except Exception as e:
            report_error(e, context={"user_id": user_id}, request=request)
            raise  # re-raise after reporting
    """

    def __init__(self, endpoint: str | None = None, service_name: str | None = None,
                 environment: str | None = None) -> None:
        self.endpoint = endpoint or os.getenv("AUDIT_SERVICE_URL", "http://audit-service:8006")
        self.service_name = service_name or os.getenv("SERVICE_NAME", "unknown")
        self.environment = environment or os.getenv("ENVIRONMENT", "development")
        self.hostname = socket.gethostname()

    def report(
        self,
        error: Exception,
        context: dict[str, Any] | None = None,
        request_id: str | None = None,
        correlation_id: str | None = None,
        user_id: str | None = None,
        level: str = "error",
    ) -> None:
        """Report an error. Best-effort, never raises."""
        try:
            payload = {
                "service": self.service_name,
                "environment": self.environment,
                "level": level,
                "error_type": type(error).__name__,
                "message": str(error),
                "stack_trace": traceback.format_exc(),
                "context": {
                    "hostname": self.hostname,
                    **(context or {}),
                },
                "request_id": request_id,
                "correlation_id": correlation_id,
                "user_id": user_id,
            }
            # Send to audit-service (fire-and-forget)
            with httpx.Client(timeout=2.0) as client:
                resp = client.post(
                    f"{self.endpoint}/api/v1/internal/errors",
                    json=payload,
                    headers={"Content-Type": "application/json"},
                )
                if resp.status_code >= 400:
                    logger.warning(
                        "failed to report error to audit-service",
                        extra={"status": resp.status_code, "body": resp.text[:200]},
                    )
        except Exception as e:
            # Never let error reporting cause another error
            logger.warning("error reporter failed", exc_info=True, extra={"error": str(e)})


# Singleton
_reporter = ErrorReporter()


def report_error(
    error: Exception,
    context: dict[str, Any] | None = None,
    request=None,
    level: str = "error",
) -> None:
    """Convenience function to report error."""
    request_id = None
    correlation_id = None
    user_id = None
    if request is not None:
        request_id = getattr(request, "headers", {}).get("X-Request-Id")
        correlation_id = getattr(request, "headers", {}).get("X-Correlation-Id")
        # Try to extract user_id from request state if available
        if hasattr(request, "state"):
            user_id = getattr(request.state, "user_id", None)
    _reporter.report(error, context, request_id, correlation_id, user_id, level)

File: shared/go/observability/error_reporter.go
// Package observability provides shared error reporting for Go services.
package observability

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"runtime"
	"runtime/debug"
	"time"
)

// ErrorReporter pushes errors to the centralized audit-service.
type ErrorReporter struct {
	endpoint    string
	serviceName string
	environment string
	hostname    string
	client      *http.Client
}

// New creates an ErrorReporter from environment variables.
func New() *ErrorReporter {
	endpoint := os.Getenv("AUDIT_SERVICE_URL")
	if endpoint == "" {
		endpoint = "http://audit-service:8006"
	}
	svc := os.Getenv("SERVICE_NAME")
	if svc == "" {
		svc = "unknown"
	}
	env := os.Getenv("ENVIRONMENT")
	if env == "" {
		env = "development"
	}
	hostname, _ := os.Hostname()
	return &ErrorReporter{
		endpoint:    endpoint,
		serviceName: svc,
		environment: env,
		hostname:    hostname,
		client:      &http.Client{Timeout: 2 * time.Second},
	}
}

// Report sends an error to audit-service.
func (r *ErrorReporter) Report(
	ctx context.Context,
	err any,
	context map[string]any,
	requestID string,
	correlationID string,
	userID string,
) {
	if context == nil {
		context = map[string]any{}
	}
	context["hostname"] = r.hostname
	context["goroutines"] = runtime.NumGoroutine()

	payload := map[string]any{
		"service":        r.serviceName,
		"environment":    r.environment,
		"level":          "error",
		"error_type":     fmt.Sprintf("%T", err),
		"message":        fmt.Sprintf("%v", err),
		"stack_trace":    string(debug.Stack()),
		"context":        context,
		"request_id":     requestID,
		"correlation_id": correlationID,
		"user_id":        userID,
	}

	body, _ := json.Marshal(payload)
	req, _ := http.NewRequestWithContext(ctx, "POST",
		r.endpoint+"/api/v1/internal/errors", bytes.NewReader(body))
	req.Header.Set("Content-Type", "application/json")

	resp, err := r.client.Do(req)
	if err != nil {
		return
	}
	resp.Body.Close()
}

var defaultReporter = New()

// ReportError is a convenience function using the default reporter.
func ReportError(ctx context.Context, err any, context map[string]any) {
	defaultReporter.Report(ctx, err, context, "", "", "")
}

// ReportErrorWithRequest reports with request context.
func ReportErrorWithRequest(ctx context.Context, err any, context map[string]any,
	requestID, correlationID, userID string) {
	defaultReporter.Report(ctx, err, context, requestID, correlationID, userID)
}
